File: wf-api-integration/references/account-inquiry/notify-balance-change/java/client/NotifyBalanceChangeHandler.java
/**
 * Alipay.com Inc.
 * Copyright (c) 2004-2026 All Rights Reserved.
 */
package {basePackage}.wf.client.account;

import com.alibaba.common.logging.Logger;
import com.alibaba.common.logging.LoggerFactory;
import com.alibaba.fastjson.JSON;
import {basePackage}.wf.model.domain.BalanceChangeLog;
import {basePackage}.wf.model.domain.Amount;
import {basePackage}.wf.model.request.NotifyBalanceChangeRequest;
import {basePackage}.wf.model.response.NotifyBalanceChangeResponse;
import {basePackage}.wf.signer.WfSigner;

import java.util.List;

/**
 * WorldFirst notifyBalanceChange 回调通知处理器。
 *
 * <p>当集成商的万里汇余额账户发生动账交易后，万里汇会调用 notifyBalanceChange 接口通知集成商。
 * 本处理器负责：
 * <ol>
 *   <li>验签：校验请求签名，确保请求来自万里汇</li>
 *   <li>解析：将请求体 JSON 解析为 {@link NotifyBalanceChangeRequest}</li>
 *   <li>幂等校验：基于 {@code notifySequence} 和 {@code accountingBizNo} 进行去重</li>
 *   <li>业务处理：遍历 {@code balanceChangeLogs} 逐条处理余额变动</li>
 *   <li>构建响应：返回 {@link NotifyBalanceChangeResponse} 确认收到通知</li>
 * </ol>
 *
 * <p>若不返回成功响应，万里汇将按以下间隔重试（最多 7 次）：
 * 2分钟、10分钟、10分钟、1小时、2小时、6小时、15小时
 *
 * @author Qoder
 * @version NotifyBalanceChangeHandler.java, v 0.1 2026-04-21
 */
public class NotifyBalanceChangeHandler {

    private static final Logger LOGGER = LoggerFactory.getLogger(NotifyBalanceChangeHandler.class);

    private WfSigner signer;

    /**
     * 处理 notifyBalanceChange 回调通知。
     *
     * <p>处理流程：
     * <ol>
     *   <li>验签（调用方应在 Controller 层完成验签，此处仅做业务处理）</li>
     *   <li>解析请求体</li>
     *   <li>基于 notifySequence 做幂等校验</li>
     *   <li>遍历 balanceChangeLogs 处理每条余额变动</li>
     *   <li>返回成功响应</li>
     * </ol>
     *
     * @param requestBody 请求体 JSON 字符串
     * @return 响应对象
     */
    public NotifyBalanceChangeResponse handleNotification(String requestBody) {
        LOGGER.info("NotifyBalanceChangeHandler received notification");

        // 1. 解析请求体
        NotifyBalanceChangeRequest request;
        try {
            request = JSON.parseObject(requestBody, NotifyBalanceChangeRequest.class);
        } catch (Exception e) {
            LOGGER.error("NotifyBalanceChangeHandler failed to parse request body, body=" + requestBody, e);
            return NotifyBalanceChangeResponse.fail("UNKNOWN_EXCEPTION",
                "Failed to parse request body: " + e.getMessage());
        }

        if (request == null) {
            LOGGER.error("NotifyBalanceChangeHandler request is null");
            return NotifyBalanceChangeResponse.fail("UNKNOWN_EXCEPTION", "request body is empty");
        }

        Integer notifySequence = request.getNotifySequence();
        LOGGER.info("NotifyBalanceChangeHandler processing notifySequence=" + notifySequence);

        // 2. 幂等校验（集成商需自行实现去重逻辑）
        // TODO: 根据 notifySequence 查询是否已处理过，若已处理则直接返回成功
        // if (isDuplicate(notifySequence)) {
        //     LOGGER.info("NotifyBalanceChangeHandler duplicate notifySequence=" + notifySequence + ", skip processing");
        //     return NotifyBalanceChangeResponse.success();
        // }

        // 3. 遍历 balanceChangeLogs 处理每条余额变动
        List<BalanceChangeLog> changeLogs = request.getBalanceChangeLogs();
        if (changeLogs == null || changeLogs.isEmpty()) {
            LOGGER.warn("NotifyBalanceChangeHandler balanceChangeLogs is empty, notifySequence=" + notifySequence);
            return NotifyBalanceChangeResponse.success();
        }

        LOGGER.info("NotifyBalanceChangeHandler processing " + changeLogs.size()
            + " balance change log(s), notifySequence=" + notifySequence);

        for (BalanceChangeLog changeLog : changeLogs) {
            try {
                processBalanceChangeLog(changeLog);
            } catch (Exception e) {
                LOGGER.error("NotifyBalanceChangeHandler failed to process changeLog, accountingBizNo="
                    + (changeLog != null ? changeLog.getAccountingBizNo() : "null"), e);
                return NotifyBalanceChangeResponse.fail("UNKNOWN_EXCEPTION",
                    "Failed to process balance change log: " + e.getMessage());
            }
        }

        // 4. 标记已处理（集成商需自行实现持久化逻辑）
        // TODO: markAsProcessed(notifySequence);

        // 5. 返回成功响应
        LOGGER.info("NotifyBalanceChangeHandler processed successfully, notifySequence=" + notifySequence);
        return NotifyBalanceChangeResponse.success();
    }

    /**
     * 处理单条余额变动记录。
     *
     * <p>集成商需在此方法中实现具体的业务逻辑，例如：
     * <ul>
     *   <li>根据 transactionType 分发到不同的业务处理逻辑</li>
     *   <li>更新账户余额</li>
     *   <li>记录交易流水</li>
     *   <li>发送内部通知</li>
     * </ul>
     *
     * @param changeLog 余额变动记录
     */
    private void processBalanceChangeLog(BalanceChangeLog changeLog) {
        String accountingBizNo = changeLog.getAccountingBizNo();
        String transactionType = changeLog.getTransactionType();
        String accountNo = changeLog.getAccountNo();
        Amount transactionAmount = changeLog.getTransactionAmount();
        Amount accountBalance = changeLog.getAccountBalance();

        LOGGER.info("NotifyBalanceChangeHandler processing changeLog:"
            + " accountingBizNo=" + accountingBizNo
            + ", transactionType=" + transactionType
            + ", accountNo=" + accountNo
            + ", transactionAmount=" + (transactionAmount != null
                ? transactionAmount.getValue() + " " + transactionAmount.getCurrency() : "null")
            + ", accountBalance=" + (accountBalance != null
                ? accountBalance.getValue() + " " + accountBalance.getCurrency() : "null")
            + ", balanceType=" + changeLog.getBalanceType()
            + ", balanceChangeTime=" + changeLog.getBalanceChangeTime());

        // TODO: 集成商根据 transactionType 实现具体业务逻辑
        // switch (transactionType) {
        //     case "TRANSFER":
        //         handleTransfer(changeLog);
        //         break;
        //     case "TRANSFER_REFUND":
        //         handleTransferRefund(changeLog);
        //         break;
        //     case "COLLECTION":
        //         handleCollection(changeLog);
        //         break;
        //     case "WITHDRAWAL":
        //         handleWithdrawal(changeLog);
        //         break;
        //     default:
        //         LOGGER.info("NotifyBalanceChangeHandler unhandled transactionType=" + transactionType);
        //         break;
        // }
    }

    // ==================== Getter/Setter ====================

    /**
     * Setter method for property <tt>signer</tt>.
     *
     * @param signer value to be assigned to property signer
     */
    public void setSigner(WfSigner signer) {
        this.signer = signer;
    }

    /**
     * Getter method for property <tt>signer</tt>.
     *
     * @return property value of signer
     */
    public WfSigner getSigner() {
        return signer;
    }
}

File: wf-api-integration/references/account-inquiry/notify-balance-change/golang/client/notify_balance_change_handler.go
package client

import (
	"encoding/json"
	"fmt"
	"log"

	"{moduleName}/wf/model/request"
	"{moduleName}/wf/model/response"
)

// NotifyBalanceChangeHandler handles notifyBalanceChange callback notifications from WF.
//
// When a transaction occurs in the integrator's WF balance account, WF sends a
// notifyBalanceChange callback. This handler is responsible for:
//  1. Parsing the request body
//  2. Idempotency check based on notifySequence
//  3. Processing each balance change log
//  4. Building the response to acknowledge receipt
//
// If no successful response is returned, WF will retry up to 7 times with intervals:
// 2min, 10min, 10min, 1h, 2h, 6h, 15h
type NotifyBalanceChangeHandler struct {
	// OnBalanceChange is called for each balance change log in the notification.
	// Integrators should implement this callback to handle business logic
	// (e.g., update balance, record transaction, send internal notification).
	OnBalanceChange func(changeLog *request.BalanceChangeLog)
}

// NewNotifyBalanceChangeHandler creates a new NotifyBalanceChangeHandler with default logging callback.
func NewNotifyBalanceChangeHandler() *NotifyBalanceChangeHandler {
	return &NotifyBalanceChangeHandler{
		OnBalanceChange: func(changeLog *request.BalanceChangeLog) {
			log.Printf("[NotifyBalanceChangeHandler] balance change: accountingBizNo=%s, transactionType=%s, "+
				"accountNo=%s, amount=%d %s, balance=%d %s",
				changeLog.AccountingBizNo, changeLog.TransactionType,
				changeLog.AccountNo,
				changeLog.TransactionAmount.Value, changeLog.TransactionAmount.Currency,
				changeLog.AccountBalance.Value, changeLog.AccountBalance.Currency)
		},
	}
}

// HandleNotification processes a notifyBalanceChange callback notification.
//
// The caller (e.g., HTTP handler) should:
//  1. Verify the request signature before calling this method
//  2. Pass the raw request body as []byte
//  3. Marshal the returned response and send it back to WF
//
// Returns a NotifyBalanceChangeResponse that should be sent back to WF.
func (h *NotifyBalanceChangeHandler) HandleNotification(body []byte) *response.NotifyBalanceChangeResponse {
	log.Println("[NotifyBalanceChangeHandler] received notification")

	// 1. Parse request body
	var req request.NotifyBalanceChangeRequest
	if err := json.Unmarshal(body, &req); err != nil {
		log.Printf("[NotifyBalanceChangeHandler] failed to parse request body: %v", err)
		return response.NewFailResponse("UNKNOWN_EXCEPTION",
			fmt.Sprintf("failed to parse request body: %v", err))
	}

	log.Printf("[NotifyBalanceChangeHandler] processing notifySequence=%d", req.NotifySequence)

	// 2. Idempotency check (integrators should implement their own deduplication logic)
	// TODO: Check if notifySequence has been processed before
	// if isDuplicate(req.NotifySequence) {
	//     log.Printf("[NotifyBalanceChangeHandler] duplicate notifySequence=%d, skip processing", req.NotifySequence)
	//     return response.NewSuccessResponse()
	// }

	// 3. Process each balance change log
	if len(req.BalanceChangeLogs) == 0 {
		log.Printf("[NotifyBalanceChangeHandler] balanceChangeLogs is empty, notifySequence=%d", req.NotifySequence)
		return response.NewSuccessResponse()
	}

	log.Printf("[NotifyBalanceChangeHandler] processing %d balance change log(s), notifySequence=%d",
		len(req.BalanceChangeLogs), req.NotifySequence)

	for i := range req.BalanceChangeLogs {
		changeLog := &req.BalanceChangeLogs[i]
		if h.OnBalanceChange != nil {
			h.OnBalanceChange(changeLog)
		}
	}

	// 4. Mark as processed (integrators should implement their own persistence logic)
	// TODO: markAsProcessed(req.NotifySequence)

	// 5. Return success response
	log.Printf("[NotifyBalanceChangeHandler] processed successfully, notifySequence=%d", req.NotifySequence)
	return response.NewSuccessResponse()
}
